### ISSUE TITLE
Unpacking variable sized payloads.

### ISSUE THREAD

**Issue by mhussa** [go-restruct/restruct#26]
Hello, Thankyou for creating restruct.  I've had a tinker and can get a few basic things working but I have a problem unpacking nested messages.  I get a "runtime error:  slice bounds out of range" when I try to Unpack a buffer using these structures.

````
type XDPMsg struct {
    Size uint16 `struct:"uint16,little,sizeof=Data"`
    Type uint16 `struct:"uint16,little"`
    Data []uint8 
  }

type NyseXDPPacket struct {
  Size uint `struct:"uint16,little"`
  DeliveryFlag uint `struct:"uint8"`
  MessageCount uint8 `struct:"uint8,sizeof=Msgs"`
  SequenceNumber uint `struct:"uint32,little"`
  SendTime uint `struct:"uint32,little"`
  SendTimeNS uint `struct:"uint32,little"`
  Msgs []XDPMsg
}
````

I performed some debugging and discovered that when I Unpack a buffer using the above structure definitions, the Data[] of the first Msg is created correctly but the next message fails because restruc seems to be not skipping the Data[] bytes of the first message.  It reads the Size and Type from the next 4 bytes after the first Msg Size and Type and tries to create an out of range slice here:

````
case reflect.Slice:
              sv.Set(reflect.MakeSlice(sf.Type, sl, sl))
````

Is there some tag I need to add "Data[]uint8" to make this work?

Any help appreciated.

**Comment by mhussa** [comment]
I think this is the bit of logic in field.go causing the behaviour I'm seeing.


                // Optimization: if the array/slice is empty, bail now.
                if alen == 0 {
                        return size
                }

On creating the "Msgs" array this is what I want it do, return "size", but in the "Data[]" case I want it to return the value in "Size".  I may have to tinker with the tag types to make it do what I need.  Skip won't work because then the data doesn't get unpacked.

**Comment by mhussa** [comment]
The problem was that the "Size" field value is inclusive of the entire structure not just the Data payload size.  restruc (sensibly) uses the size to determine the size of Data but it ends up consuming 4 bytes more per payload and goes out of sync reading out of bounds lengths.

So... 

I added an "adjust" tag which subtracts 4 from the value it reads into the size field.  This makes sure the right amount of data is decoded.  It works only for decoding,  If I packed the data using this it would be incorrect.  Not a great solution but I'm only interested in a decode at the moment.

Hope this helps anyone else comes across this.
